### PR TITLE
Refactor Canary to follow schema of remaining ASR models

### PR DESCRIPTION
**Pull request by titu1994**
# What does this PR do ?

Refactor Canary to follow Archetype pattern of rest of ASR. Add Decoding Strategy to disentangle decoding with model  architecture.

**Collection**: [ASR]

# Changelog 
- Refactors model class to conform to ASR's archetype pattern
- Add Decoding strategy as a middle layer between model logits and text to allow extending decoding algorithms in the future
- Update configs to new schema

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 1]
```diff
@@ -180,28 +190,42 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
+        with open_dict(self.cfg.decoding):
+            self.cfg.decoding = decoding_cfg
+
+        logging.info(f"Changed decoding strategy to \n{OmegaConf.to_yaml(self.cfg.decoding)}")
+
     @torch.no_grad()
     def transcribe(
         self,
-        paths2audio_files: Union[List[str], str],
+        audio: Union[List[str], str],
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:217]: Well eventually merge this transcription with the new transcription API - it changes this name to `audio` and deprecates and removes `logprobs`.

File: nemo/collections/asr/parts/submodules/multitask_beam_decoding.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    #     return {
+    #         "decoder_output": NeuralType(None, LogprobsType()),
+    #         "decoder_lengths": NeuralType(tuple('B'), LengthsType()),
+    #     }
+    #
+    # @property
+    # def output_types(self):
+    #     """Returns definitions of module output ports.
+    #     """
+    #     return {"predictions": [NeuralType(elements_type=HypothesisType())]}
```
Comment (pzelasko) [inline nemo/collections/asr/parts/submodules/multitask_beam_decoding.py:132]: Remove/restore commeted out code

File: nemo/collections/asr/parts/submodules/multitask_decoding.py [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+
+                max_generation_delta: int, maximum number of additional target tokens to generate
+
+                return_best_hypothesis: optional bool, whether to return just the best hypothesis or all of the
+                    hypotheses after beam search has concluded. This flag is set by default.
+
+
+        transformer_decoder: Transformer decoder module.
+        log_softmax_module: Log Softmax projection module to the vocab size.
+        tokenizer: Aggregate Tokenizer.
```
Comment (pzelasko) [inline nemo/collections/asr/parts/submodules/multitask_decoding.py:87]: Does it have to be an aggregate tokenizer? the typing below specifies `TokenizerSpec`. We might want to use non-aggregate tokenizer later

File: nemo/collections/asr/parts/submodules/multitask_decoding.py [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+                    prediction = hypotheses_list[ind].y_sequence
+
+                    if type(prediction) != list:
+                        prediction = prediction.tolist()
+
+                    hypotheses[ind].langs = self.decode_tokens_to_lang(prediction)
+                    hypotheses[ind].langs_chars = self.decode_ids_to_langs(prediction)
+            else:
+                logging.warning(
+                    "Ignoring request for lang output in hypotheses since the model does not use an aggregate tokenizer"
```
Comment (pzelasko) [inline nemo/collections/asr/parts/submodules/multitask_decoding.py:466]: should we just require aggregate tokenizer for this version and if we support non-aggregate tokenizer later we would write another subclass for it?

File: nemo/collections/asr/parts/submodules/multitask_decoding.py [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+    # preserve decoding alignments
+    preserve_alignments: Optional[bool] = None
+
+    # compute language IDs
+    compute_langs: bool = False
+
+    # greedy decoding config
+    # greedy: rnnt_greedy_decoding.GreedyBatchedRNNTInferConfig = field(
+    #     default_factory=rnnt_greedy_decoding.GreedyBatchedRNNTInferConfig
+    # )
```
Comment (pzelasko) [inline nemo/collections/asr/parts/submodules/multitask_decoding.py:487]: Remove/restore commented out code (rnnt probably doesn't belong here?)

**Review by pzelasko — APPROVED** [review]
Left a couple of comments/questions. Great work, LGTM!

File: nemo/collections/asr/parts/submodules/multitask_beam_decoding.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/multitask_beam_decoding.py:1]: update year.

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 7]
```diff
@@ -95,83 +95,83 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
-        self.transf_loss = SmoothedCrossEntropyLoss(
-            pad_id=self.tokenizer.pad_id, label_smoothing=self.cfg.label_smoothing
-        )
+        with open_dict(self.cfg.loss):
+            self.cfg.loss.pad_id = self.tokenizer.pad_id
+
+        self.loss = EncDecMultiTaskModel.from_config_dict(self.cfg.loss)
+        # self.transf_loss = SmoothedCrossEntropyLoss(
+        #     pad_id=self.tokenizer.pad_id, label_smoothing=self.cfg.label_smoothing
+        # )
```
Comment (nithinraok) [inline nemo/collections/asr/models/aed_multitask_models.py:174]: clean these comments.

File: nemo/collections/asr/parts/submodules/multitask_decoding.py [hunk 8]
```diff
@@ -0,0 +1,493 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (nithinraok) [inline nemo/collections/asr/parts/submodules/multitask_decoding.py:1]: update year

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 9]
```diff
@@ -95,83 +95,82 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
+
+        # Setup token classifier
+        with open_dict(self.cfg.head):
+            self.cfg.head.num_classes = vocab_size
+
+        self.log_softmax = EncDecMultiTaskModel.from_config_dict(self.cfg.head)
+        if isinstance(self.log_softmax, MultiLayerPerceptron):
+            self.log_softmax.layer0.weight = self.transf_decoder.embedding.token_embedding.weight
+        else:
+            raise ValueError(f"Expected MultiLayerPerceptron head, got {type(self.log_softmax)}")
```
Comment (stevehuang52) [inline nemo/collections/asr/models/aed_multitask_models.py:144]: Shall we force the classifier to `MultiLayerPerceptron`? Putting it in config may caused errors by the users. @titu1994

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 10]
```diff
@@ -52,15 +52,11 @@
...
 try:
     from sacrebleu import corpus_bleu
 
-    from nemo.collections.nlp.modules.common import TokenClassifier
-    from nemo.collections.nlp.modules.common.lm_utils import get_transformer
-    from nemo.collections.nlp.modules.common.transformer import BeamSearchSequenceGenerator, TransformerEncoder
-
     NLP_AVAILABLE = True
 except (ImportError, ModuleNotFoundError):
     NLP_AVAILABLE = False
```
Comment (krishnacpuvvada) [inline nemo/collections/asr/models/aed_multitask_models.py:62]: remove this NLP import check after we switch over to ASR only dependency @stevehuang52

Comment (stevehuang52) [reply nemo/collections/asr/models/aed_multitask_models.py:63]: done

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 11]
```diff
@@ -95,83 +96,81 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
+
+        # Setup token classifier
+        with open_dict(self.cfg.head):
+            self.cfg.head.num_classes = vocab_size
+
+        self.log_softmax = EncDecMultiTaskModel.from_config_dict(self.cfg.head)
+        if isinstance(self.log_softmax, TokenClassifier):
+            self.log_softmax.mlp.layer0.weight = self.transf_decoder.embedding.token_embedding.weight
+        else:
+            raise ValueError(f"Expected TokenClassifier head, got {type(self.log_softmax)}")
```
Comment (stevehuang52) [inline nemo/collections/asr/models/aed_multitask_models.py:145]: Shall we fix the class for self.log_softmax to be TokenClassifier? @titu1994

File: nemo/collections/asr/parts/submodules/multitask_beam_decoding.py (hunk 2 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/multitask_beam_decoding.py:118]: This is intentional - we need to add neural types to all the stuff we're adding right now, just that it's lower priority. There's a few to-dos sprinkled all over the PR

File: nemo/collections/asr/parts/submodules/multitask_decoding.py (hunk 3 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/multitask_decoding.py:86]: Hmm good point

File: nemo/collections/asr/parts/submodules/multitask_decoding.py (hunk 5 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/multitask_decoding.py:486]: This is placeholder for @AlexGrinch to add greedy decoding strategy

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 11 above)
Comment (titu1994) [reply nemo/collections/asr/models/aed_multitask_models.py:145]: Hmm. I suppose that's one way but for now let's keep it as is. There's a lot of places that have hard coded checks like this, it violates abstraction principles. Rigit now there's no time to fix this but we need to improve code later 

So let's keep it as is for now, and hope users don't change token classifier

Comment (titu1994) [reply nemo/collections/asr/models/aed_multitask_models.py:145]: Actually instead of doing this weight tying - what's we can do is ignore it if TokenClasskfier is not used. What do you think @AlexGrinch 

Nlp has stopped doing weight tying anyway

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml [hunk 12]
```diff
@@ -132,45 +135,70 @@ model:
...
+    encoder: null
     pre_ln_final_layer_norm: true
 
+    config_dict:
+      max_sequence_length: 512
+      num_token_types: 0
+      embedding_dropout: 0.1
+      learn_positional_encodings: false
+      hidden_size: 1024
+      inner_size: 4096
```
Comment (krishnacpuvvada) [inline examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:166]: as we are refactoring, can we change this spec to more intuitive -  (hidden_size, ff_expansion_factor) like conformer rather than (hidden_size, inner_size),

Comment (titu1994) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:166]: That's more on model side plus it's common nomenclature for transformers ?

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 13]
```diff
@@ -180,28 +179,42 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
-            max_sequence_length=self.transf_decoder.max_sequence_length,
-            beam_size=cfg.beam_size,
-            bos=self.tokenizer.bos_id,
-            pad=self.tokenizer.pad_id,
-            eos=self.tokenizer.eos_id,
-            len_pen=cfg.len_pen,
-            max_delta_length=cfg.max_generation_delta,
+    def change_decoding_strategy(self, decoding_cfg: DictConfig):
+        """
+        Changes decoding strategy used during CTC decoding process.
```
Comment (krishnacpuvvada) [inline nemo/collections/asr/models/aed_multitask_models.py:184]: remove CTC reference ? :)

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 14]
```diff
@@ -226,26 +238,24 @@ def transcribe(
...
+        if audio is None or len(audio) == 0:
             return {}
 
         if return_hypotheses:
             logging.warning("return_hypotheses=True is currently not supported, returning text instead.")
 
-        if isinstance(paths2audio_files, list):
-            logging.info(f"Found paths2audio_files to be a list of {len(paths2audio_files)} items.")
+        if isinstance(audio, list):
+            logging.info(f"Found paths2audio_files to be a list of {len(audio)} items.")
```
Comment (krishnacpuvvada) [inline nemo/collections/asr/models/aed_multitask_models.py:248]: remove references to `paths2audio_files` now that it is changed to `audio`

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 15]
```diff
@@ -226,26 +238,24 @@ def transcribe(
...
             logging.info(f"Assuming each item in paths2audio_files is a path to audio file.")
             logging.info(f"Transcribing with default Canary setting of English without PnC.")
```
Comment (krishnacpuvvada) [inline nemo/collections/asr/models/aed_multitask_models.py:250]: remove Canary reference ?

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 16]
```diff
@@ -95,83 +96,81 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
+        self.context_len_for_AR_decoding = self.cfg.get("context_len_for_AR_decoding", 5)
 
         # Define autoregressive CE loss
-        self.transf_loss = SmoothedCrossEntropyLoss(
-            pad_id=self.tokenizer.pad_id, label_smoothing=self.cfg.label_smoothing
-        )
+        with open_dict(self.cfg.loss):
+            self.cfg.loss.pad_id = self.tokenizer.pad_id
+
+        self.loss = EncDecMultiTaskModel.from_config_dict(self.cfg.loss)
```
Comment (krishnacpuvvada) [inline nemo/collections/asr/models/aed_multitask_models.py:173]: any reason why we removed `label_smoothing` here ? yaml config has `label_smoothing` parameter.

File: nemo/collections/asr/parts/submodules/multitask_decoding.py (hunk 4 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/multitask_decoding.py:465]: Langid can only be retrieved by agg tokenizer, but its not a necessity for other tokenizers

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 7 above)
Comment (titu1994) [reply nemo/collections/asr/models/aed_multitask_models.py]: Done

File: nemo/collections/asr/parts/submodules/multitask_decoding.py [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+        Args:
+            tokens: List of int representing the token ids.
+
+        Returns:
+            A list of decoded tokens.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def decode_tokens_to_lang(self, tokens: List[int]) -> str:
```
Comment (krishnacpuvvada) [inline nemo/collections/asr/parts/submodules/multitask_decoding.py:274]: Looks specific to aggregate tokenizer. should we include this in abstract class.

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml (hunk 12 above)
Comment (titu1994) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:166]: Made the config more adaptive

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 15 above)
Comment (titu1994) [reply nemo/collections/asr/models/aed_multitask_models.py:250]: Made it check if tokenizer is agg tokenizer, and then print first language name. Removed the stuff for PC - cant guarantee that tokenizer has support for PC or not.

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 16 above)
Comment (titu1994) [reply nemo/collections/asr/models/aed_multitask_models.py:173]: Forgot. Changed the config to read it from the top of the config with

```yaml
  loss:
    _target_: nemo.collections.common.losses.smoothed_cross_entropy.SmoothedCrossEntropyLoss
    label_smoothing: ${model.label_smoothing}
    pad_id: null
```

File: nemo/collections/asr/parts/submodules/multitask_decoding.py (hunk 17 above)
Comment (titu1994) [reply nemo/collections/asr/parts/submodules/multitask_decoding.py:274]: I agree, but RNNT has similar API for some reason. Well it would be necessary for other tokenizers to also use this function for language calculation, and it only gets called if used explicitly so it might be ok for now. But i do agree this is too specific to be part of abstract class. Fyi @bmwshop @KunalDhawan

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml (hunk 12 above)
Comment (krishnacpuvvada) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:166]: you are right Pytorch Transformer uses `d_model` and `dim_feedforward`..

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml [hunk 18]
```diff
@@ -132,45 +135,70 @@ model:
...
-    num_attention_heads: 8
-    ffn_dropout: 0.1
-    attn_score_dropout: 0.1
-    attn_layer_dropout: 0.1
-    hidden_act: relu
-    pre_ln: true
+    encoder: null
     pre_ln_final_layer_norm: true
 
+    config_dict:
```
Comment (krishnacpuvvada) [inline examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:160]: we are using `config_dict` for `transf_decoder` but not for `transf_encoder` (i see the reason why it is way in model init.. so just a note )

Comment (titu1994) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:169]: Yeah the get_transformers function is really sad but its not worth refactoring the function. This is fine for the time being i think..

File: nemo/core/classes/modelPT.py [hunk 19]
```diff
@@ -42,6 +42,10 @@
 __all__ = ['ModelPT']
 
 
+# multiple interpolated values in the config
+OmegaConf.register_new_resolver("multiply", lambda x, y: x * y, replace=True)
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:46]: @ericharper heads up. ASR kinda needs the ability for Hydra multiply as well, so I'm making it a mandatory resolver by adding it into the ModelPT class (since all nemo models import it, omegaconf in NeMo usage will always support multiply op).

Keeping the support inside of hydra_runner as well - just making it so it wont crash due to already registered component.

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 13 above)
Comment (titu1994) [reply nemo/collections/asr/models/aed_multitask_models.py]: Done

File: nemo/collections/asr/models/aed_multitask_models.py (hunk 14 above)
Comment (titu1994) [reply nemo/collections/asr/models/aed_multitask_models.py:253]: Ah docstring. We'll have to do a major fix later anyway
